File: src/ekore/anomalous_dimensions/unpolarized/space_like/aem2.py
"""The :math:`O(a_{em}^2)` Altarelli-Parisi splitting kernels."""

import numba as nb
import numpy as np

from eko import constants

from ....harmonics import cache as c
from . import as1aem1


@nb.njit(cache=True)
def gamma_phph(N, nf):
    r"""Compute the :math:`O(a_{em}^2)` photon-photon singlet anomalous
    dimension.

    Implements Eq. (68) of :cite:`deFlorian:2016gvk`.

    Parameters
    ----------
    N : complex
        Mellin moment
    nf : int
        Number of active flavors

    Returns
    -------
    complex
        :math:`O(a_{em}^2)` photon-photon singlet anomalous dimension
        :math:`\\gamma_{\\gamma \\gamma}^{(0,2)}(N)`
    """
    nu = constants.uplike_flavors(nf)
    nd = nf - nu
    return (
        constants.NC
        * (nu * constants.eu2**2 + nd * constants.ed2**2)
        * (as1aem1.gamma_gph(N) / constants.CF / constants.CA + 4)
    )


@nb.njit(cache=True)
def gamma_uph(N, nf, cache):
    r"""Compute the :math:`O(a_{em}^2)` quark-photon anomalous dimension for up
    quarks.

    Implements Eq. (55) of :cite:`deFlorian:2016gvk` for q=u.

    Parameters
    ----------
    N : complex
        Mellin moment
    nf : int
        Number of active flavors
    cache: numpy.ndarray
        Harmonic sum cache

    Returns
    -------
    complex
        :math:`O(a_{em}^2)` quark-photon anomalous dimension :math:`\\gamma_{u \\gamma}^{(0,2)}(N)`
    """
    return constants.eu2 * as1aem1.gamma_qph(N, nf, cache) / constants.CF


@nb.njit(cache=True)
def gamma_dph(N, nf, cache):
    r"""Compute the :math:`O(a_{em}^2)` quark-photon anomalous dimension for
    down quarks.

    Implements Eq. (55) of :cite:`deFlorian:2016gvk` for q=d.

    Parameters
    ----------
    N : complex
        Mellin moment
    nf : int
        Number of active flavors
    cache: numpy.ndarray
        Harmonic sum cache

    Returns
    -------
    complex
        :math:`O(a_{em}^2)` quark-photon anomalous dimension :math:`\\gamma_{d \\gamma}^{(0,2)}(N)`
    """
    return constants.ed2 * as1aem1.gamma_qph(N, nf, cache) / constants.CF


@nb.njit(cache=True)
def gamma_phu(N, nf, cache):
    r"""Compute the :math:`O(a_{em}^2)` photon-quark anomalous dimension for up
    quarks.

    Implements Eq. (56) of :cite:`deFlorian:2016gvk` for q=u.

    Parameters
    ----------
    N : complex
        Mellin moment
    nf : int
        Number of active flavors
    cache: numpy.ndarray
        Harmonic sum cache

    Returns
    -------
    complex
        :math:`O(a_{em}^2)` photon-quark anomalous dimension :math:`\\gamma_{\\gamma u}^{(0,2)}(N)`
    """
    nu = constants.uplike_flavors(nf)
    nd = nf - nu
    S1 = c.get(c.S1, cache, N)
    tmp = (-16.0 * (-16.0 - 27.0 * N - 13.0 * N**2 - 8.0 * N**3)) / (
        9.0 * (-1.0 + N) * N * (1.0 + N) ** 2
    ) - 16.0 * (2.0 + 3.0 * N + 2.0 * N**2 + N**3) / (
        3.0 * (-1.0 + N) * N * (1.0 + N) ** 2
    ) * S1
    eSigma2 = constants.NC * (nu * constants.eu2 + nd * constants.ed2)
    return constants.eu2 * as1aem1.gamma_phq(N, cache) / constants.CF + eSigma2 * tmp


@nb.njit(cache=True)
def gamma_phd(N, nf, cache):
    r"""Compute the :math:`O(a_{em}^2)` photon-quark anomalous dimension for
    down quarks.

    Implements Eq. (56) of :cite:`deFlorian:2016gvk` for q=d.

    Parameters
    ----------
    N : complex
        Mellin moment
    nf : int
        Number of active flavors
    cache: numpy.ndarray
        Harmonic sum cache

    Returns
    -------
    complex
        :math:`O(a_{em}^2)` photon-quark anomalous dimension :math:`\\gamma_{\\gamma d}^{(0,2)}(N)`
    """
    nu = constants.uplike_flavors(nf)
    nd = nf - nu
    S1 = c.get(c.S1, cache, N)
    tmp = (-16.0 * (-16.0 - 27.0 * N - 13.0 * N**2 - 8.0 * N**3)) / (
        9.0 * (-1.0 + N) * N * (1.0 + N) ** 2
    ) - 16.0 * (2.0 + 3.0 * N + 2.0 * N**2 + N**3) / (
        3.0 * (-1.0 + N) * N * (1.0 + N) ** 2
    ) * S1
    eSigma2 = constants.NC * (nu * constants.eu2 + nd * constants.ed2)
    return constants.ed2 * as1aem1.gamma_phq(N, cache) / constants.CF + eSigma2 * tmp


@nb.njit(cache=True)
def gamma_nspu(N, nf, cache):
    r"""Compute the :math:`O(a_{em}^2)` singlet-like non-singlet anomalous
    dimension for up quarks.

    Implements sum of Eqs. (57-58) of :cite:`deFlorian:2016gvk` for q=u.

    Parameters
    ----------
    N : complex
        Mellin moment
    nf : int
        Number of active flavors
    cache: numpy.ndarray
        Harmonic sum cache

    Returns
    -------
    complex
        :math:`O(a_{em}^2)` singlet-like non-singlet anomalous dimension
        :math:`\\gamma_{ns,+,u}^{(0,2)}(N)`
    """
    S1 = c.get(c.S1, cache, N)
    S2 = c.get(c.S2, cache, N)
    nu = constants.uplike_flavors(nf)
    nd = nf - nu
    eSigma2 = constants.NC * (nu * constants.eu2 + nd * constants.ed2)
    tmp = (
        2.0
        * (-12.0 + 20.0 * N + 47.0 * N**2 + 6.0 * N**3 + 3.0 * N**4)
        / (9.0 * N**2 * (1.0 + N) ** 2)
        - 80.0 / 9.0 * S1
        + 16.0 / 3.0 * S2
    ) * eSigma2
    return constants.eu2 * as1aem1.gamma_nsp(N, cache) / constants.CF / 2.0 + tmp


@nb.njit(cache=True)
def gamma_nspd(N, nf, cache):
    r"""Compute the :math:`O(a_{em}^2)` singlet-like non-singlet anomalous
    dimension for down quarks.

    Implements sum of Eqs. (57-58) of :cite:`deFlorian:2016gvk` for q=d.

    Parameters
    ----------
    N : complex
        Mellin moment
    nf : int
        Number of active flavors
    cache: numpy.ndarray
        Harmonic sum cache

    Returns
    -------
    complex
        :math:`O(a_{em}^2)` singlet-like non-singlet anomalous dimension
        :math:`\\gamma_{ns,+,d}^{(0,2)}(N)`
    """
    S1 = c.get(c.S1, cache, N)
    S2 = c.get(c.S2, cache, N)
    nu = constants.uplike_flavors(nf)
    nd = nf - nu
    eSigma2 = constants.NC * (nu * constants.eu2 + nd * constants.ed2)
    tmp = (
        2.0
        * (-12.0 + 20.0 * N + 47.0 * N**2 + 6.0 * N**3 + 3.0 * N**4)
        / (9.0 * N**2 * (1.0 + N) ** 2)
        - 80.0 / 9.0 * S1
        + 16.0 / 3.0 * S2
    ) * eSigma2
    return constants.ed2 * as1aem1.gamma_nsp(N, cache) / constants.CF / 2.0 + tmp


@nb.njit(cache=True)
def gamma_nsmu(N, nf, cache):
    r"""Compute the :math:`O(a_{em}^2)` valence-like non-singlet anomalous
    dimension for up quarks.

    Implements difference between Eqs. (57-58) of :cite:`deFlorian:2016gvk` for q=u.

    Parameters
    ----------
    N : complex
        Mellin moment
    nf : int
        Number of active flavors
    cache: numpy.ndarray
        Harmonic sum cache

    Returns
    -------
    complex
        :math:`O(a_{em}^2)` valence-like non-singlet anomalous dimension
        :math:`\\gamma_{ns,-,u}^{(0,2)}(N)`
    """
    S1 = c.get(c.S1, cache, N)
    S2 = c.get(c.S2, cache, N)
    nu = constants.uplike_flavors(nf)
    nd = nf - nu
    eSigma2 = constants.NC * (nu * constants.eu2 + nd * constants.ed2)
    tmp = (
        2.0
        * (-12.0 + 20.0 * N + 47.0 * N**2 + 6.0 * N**3 + 3.0 * N**4)
        / (9.0 * N**2 * (1.0 + N) ** 2)
        - 80.0 / 9.0 * S1
        + 16.0 / 3.0 * S2
    ) * eSigma2
    return constants.eu2 * as1aem1.gamma_nsm(N, cache) / constants.CF / 2.0 + tmp


@nb.njit(cache=True)
def gamma_nsmd(N, nf, cache):
    r"""Compute the :math:`O(a_{em}^2)` valence-like non-singlet anomalous
    dimension for down quarks.

    Implements difference between Eqs. (57-58) of :cite:`deFlorian:2016gvk` for q=d.

    Parameters
    ----------
    N : complex
        Mellin moment
    nf : int
        Number of active flavors
    cache: numpy.ndarray
        Harmonic sum cache

    Returns
    -------
    complex
        :math:`O(a_{em}^2)` valence-like non-singlet anomalous dimension
        :math:`\\gamma_{ns,-,d}^{(0,2)}(N)`
    """
    S1 = c.get(c.S1, cache, N)
    S2 = c.get(c.S2, cache, N)
    nu = constants.uplike_flavors(nf)
    nd = nf - nu
    eSigma2 = constants.NC * (nu * constants.eu2 + nd * constants.ed2)
    tmp = (
        2.0
        * (-12.0 + 20.0 * N + 47.0 * N**2 + 6.0 * N**3 + 3.0 * N**4)
        / (9.0 * N**2 * (1.0 + N) ** 2)
        - 80.0 / 9.0 * S1
        + 16.0 / 3.0 * S2
    ) * eSigma2
    return constants.ed2 * as1aem1.gamma_nsm(N, cache) / constants.CF / 2.0 + tmp


@nb.njit(cache=True)
def gamma_ps(N, nf):
    r"""Compute the :math:`O(a_{em}^2)` pure-singlet quark-quark anomalous
    dimension.

    Implements Eq. (59) of :cite:`deFlorian:2016gvk`.

    Parameters
    ----------
    N : complex
        Mellin moment
    nf : int
        Number of active flavors

    Returns
    -------
    complex
        :math:`O(a_{em}^2)` pure-singlet quark-quark anomalous dimension
        :math:`\\gamma_{ps}^{(0,2)}(N)`
    """
    result = (
        -4.0
        * (2.0 + N * (5.0 + N))
        * (4.0 + N * (4.0 + N * (7.0 + 5.0 * N)))
        / ((-1.0 + N) * N**3 * (1.0 + N) ** 3 * (2.0 + N) ** 2)
    )
    return 2 * nf * constants.CA * result


@nb.njit(cache=True)
def gamma_singlet(N, nf, cache):
    r"""Compute the :math:`O(a_{em}^2)` singlet sector.

    Parameters
    ----------
    N : complex
        Mellin moment
    nf : int
        Number of active flavors
    cache: numpy.ndarray
        Harmonic sum cache

    Returns
    -------
    numpy.ndarray
        :math:`O(a_{em}^2)` singlet anomalous dimension :math:`\\gamma_{S}^{(0,2)}`
    """
    nu = constants.uplike_flavors(nf)
    nd = nf - nu
    vu = nu / nf
    vd = nd / nf
    e2m = constants.eu2 - constants.ed2
    e2avg = (nu * constants.eu2 + nd * constants.ed2) / nf
    e2m = constants.eu2 - constants.ed2
    gamma_ph_u = gamma_phu(N, nf, cache)
    gamma_ph_d = gamma_phd(N, nf, cache)
    gamma_u_ph = gamma_uph(N, nf, cache)
    gamma_d_ph = gamma_dph(N, nf, cache)
    gamma_ns_p_u = gamma_nspu(N, nf, cache)
    gamma_ns_p_d = gamma_nspd(N, nf, cache)
    gamma_pure_singlet = gamma_ps(N, nf)
    gamma_S_02 = np.array(
        [
            [0.0 + 0.0j, 0.0 + 0.0j, 0.0 + 0.0j, 0.0 + 0.0j],
            [
                0.0 + 0.0j,
                gamma_phph(N, nf),
                vu * constants.eu2 * gamma_ph_u + vd * constants.ed2 * gamma_ph_d,
                vu * (constants.eu2 * gamma_ph_u - constants.ed2 * gamma_ph_d),
            ],
            [
                0.0 + 0.0j,
                vu * constants.eu2 * gamma_u_ph + vd * constants.ed2 * gamma_d_ph,
                vu * constants.eu2 * gamma_ns_p_u
                + vd * constants.ed2 * gamma_ns_p_d
                + e2avg**2 * gamma_pure_singlet,
                vu
                * (
                    constants.eu2 * gamma_ns_p_u
                    - constants.ed2 * gamma_ns_p_d
                    + e2m * e2avg * gamma_pure_singlet
                ),
            ],
            [
                0.0 + 0.0j,
                vd * (constants.eu2 * gamma_u_ph - constants.ed2 * gamma_d_ph),
                vd
                * (
                    constants.eu2 * gamma_ns_p_u
                    - constants.ed2 * gamma_ns_p_d
                    + e2m * e2avg * gamma_pure_singlet
                ),
                vd * constants.eu2 * gamma_ns_p_u
                + vu * constants.ed2 * gamma_ns_p_d
                + vu * vd * e2m**2 * gamma_pure_singlet,
            ],
        ],
        np.complex128,
    )
    return gamma_S_02


@nb.njit(cache=True)
def gamma_valence(N, nf, cache):
    r"""Compute the :math:`O(a_{em}^2)` valence sector.

    Parameters
    ----------
    N : complex
        Mellin moment
    nf : int
        Number of active flavors
    cache: numpy.ndarray
        Harmonic sum cache

    Returns
    -------
    numpy.ndarray
        :math:`O(a_{em}^2)` valence anomalous dimension :math:`\\gamma_{V}^{(0,2)}`
    """
    nu = constants.uplike_flavors(nf)
    nd = nf - nu
    vu = nu / nf
    vd = nd / nf
    gamma_ns_m_u = gamma_nsmu(N, nf, cache)
    gamma_ns_m_d = gamma_nsmd(N, nf, cache)
    gamma_V_02 = np.array(
        [
            [
                vu * constants.eu2 * gamma_ns_m_u + vd * constants.ed2 * gamma_ns_m_d,
                vu * (constants.eu2 * gamma_ns_m_u - constants.ed2 * gamma_ns_m_d),
            ],
            [
                vd * (constants.eu2 * gamma_ns_m_u - constants.ed2 * gamma_ns_m_d),
                vd * constants.eu2 * gamma_ns_m_u + vu * constants.ed2 * gamma_ns_m_d,
            ],
        ],
        np.complex128,
    )
    return gamma_V_02

File: crates/ekore/src/anomalous_dimensions/unpolarized/spacelike/aem2.rs
//! |NLO| |QED|.
use num::complex::Complex;
use num::Zero;

use crate::constants::{ChargeCombinations, CA, CF, ED2, EU2, NC};
use crate::harmonics::cache::{Cache, K};

use crate::anomalous_dimensions::unpolarized::spacelike::as1aem1;

/// Compute the photon-photon anomalous dimension.
///
/// Implements Eq. (68) of [\[deFlorian:2016gvk\]][crate::bib::deFlorian2016gvk].
pub fn gamma_phph(c: &mut Cache, nf: u8) -> Complex<f64> {
    let cc = ChargeCombinations { nf };
    (NC as f64)
        * ((cc.nu() as f64) * EU2.powi(2) + (cc.nd() as f64) * ED2.powi(2))
        * (as1aem1::gamma_gph(c, nf) / CF / CA + 4.)
}

/// Compute the up-photon anomalous dimension.
///
/// Implements Eq. (55) of [\[deFlorian:2016gvk\]][crate::bib::deFlorian2016gvk] for q=u.
pub fn gamma_uph(c: &mut Cache, nf: u8) -> Complex<f64> {
    EU2 * as1aem1::gamma_qph(c, nf) / CF
}

/// Compute the down-photon anomalous dimension.
///
/// Implements Eq. (55) of [\[deFlorian:2016gvk\]][crate::bib::deFlorian2016gvk] for q=d.
pub fn gamma_dph(c: &mut Cache, nf: u8) -> Complex<f64> {
    ED2 * as1aem1::gamma_qph(c, nf) / CF
}

/// Compute the photon-quark anomalous dimension.
///
/// Implements singlet part of Eq. (56) of [\[deFlorian:2016gvk\]][crate::bib::deFlorian2016gvk].
fn gamma_phq(c: &mut Cache, nf: u8) -> Complex<f64> {
    let N = c.n();
    let S1 = c.get(K::S1);
    let gamma = (-16.0 * (-16.0 - 27.0 * N - 13.0 * N.powu(2) - 8.0 * N.powu(3)))
        / (9.0 * (-1.0 + N) * N * (1.0 + N).powu(2))
        - 16.0 * (2.0 + 3.0 * N + 2.0 * N.powu(2) + N.powu(3))
            / (3.0 * (-1.0 + N) * N * (1.0 + N).powu(2))
            * S1;
    let cc = ChargeCombinations { nf };
    let eSigma2 = (NC as f64) * cc.e2avg() * (nf as f64);
    eSigma2 * gamma
}

/// Compute the photon-up anomalous dimension.
///
/// Implements Eq. (56) of [\[deFlorian:2016gvk\]][crate::bib::deFlorian2016gvk].
pub fn gamma_phu(c: &mut Cache, nf: u8) -> Complex<f64> {
    EU2 * as1aem1::gamma_phq(c, nf) / CF + gamma_phq(c, nf)
}

/// Compute the photon-down anomalous dimension.
///
/// Implements Eq. (56) of [\[deFlorian:2016gvk\]][crate::bib::deFlorian2016gvk].
pub fn gamma_phd(c: &mut Cache, nf: u8) -> Complex<f64> {
    ED2 * as1aem1::gamma_phq(c, nf) / CF + gamma_phq(c, nf)
}

/// Compute the non-singlet anomalous dimension.
///
/// Implements singlet part of Eq. (57) of [\[deFlorian:2016gvk\]][crate::bib::deFlorian2016gvk].
fn gamma_nsq(c: &mut Cache, nf: u8) -> Complex<f64> {
    let N = c.n();
    let S1 = c.get(K::S1);
    let S2 = c.get(K::S2);
    let gamma = 2.0 * (-12.0 + 20.0 * N + 47.0 * N.powu(2) + 6.0 * N.powu(3) + 3.0 * N.powu(4))
        / (9.0 * N.powu(2) * (1.0 + N).powu(2))
        - 80.0 / 9.0 * S1
        + 16.0 / 3.0 * S2;
    let cc = ChargeCombinations { nf };
    let eSigma2 = (NC as f64) * cc.e2avg() * (nf as f64);
    eSigma2 * gamma
}

/// Compute the singlet-like non-singlet up anomalous dimension.
///
/// Implements sum of Eqs. (57-58) of [\[deFlorian:2016gvk\]][crate::bib::deFlorian2016gvk] for q=u.
pub fn gamma_nspu(c: &mut Cache, nf: u8) -> Complex<f64> {
    EU2 * as1aem1::gamma_nsp(c, nf) / CF / 2. + gamma_nsq(c, nf)
}

/// Compute the singlet-like non-singlet down anomalous dimension.
///
/// Implements sum of Eqs. (57-58) of [\[deFlorian:2016gvk\]][crate::bib::deFlorian2016gvk] for q=d.
pub fn gamma_nspd(c: &mut Cache, nf: u8) -> Complex<f64> {
    ED2 * as1aem1::gamma_nsp(c, nf) / CF / 2. + gamma_nsq(c, nf)
}

/// Compute the valence-like non-singlet up anomalous dimension.
///
/// Implements difference between Eqs. (57-58) of [\[deFlorian:2016gvk\]][crate::bib::deFlorian2016gvk] for q=u.
pub fn gamma_nsmu(c: &mut Cache, nf: u8) -> Complex<f64> {
    EU2 * as1aem1::gamma_nsm(c, nf) / CF / 2. + gamma_nsq(c, nf)
}

/// Compute the valence-like non-singlet down anomalous dimension.
///
/// Implements difference between Eqs. (57-58) of [\[deFlorian:2016gvk\]][crate::bib::deFlorian2016gvk] for q=d.
pub fn gamma_nsmd(c: &mut Cache, nf: u8) -> Complex<f64> {
    ED2 * as1aem1::gamma_nsm(c, nf) / CF / 2. + gamma_nsq(c, nf)
}

/// Compute the pure-singlet anomalous dimension.
///
/// Implements Eq. (59) of [\[deFlorian:2016gvk\]][crate::bib::deFlorian2016gvk].
pub fn gamma_ps(c: &mut Cache, nf: u8) -> Complex<f64> {
    let N = c.n();
    let result = -4.0 * (2.0 + N * (5.0 + N)) * (4.0 + N * (4.0 + N * (7.0 + 5.0 * N)))
        / ((-1.0 + N) * N.powu(3) * (1.0 + N).powu(3) * (2.0 + N).powu(2));
    2. * (nf as f64) * CA * result
}

/// Compute the singlet anomalous dimension matrix.
pub fn gamma_singlet(c: &mut Cache, nf: u8) -> [[Complex<f64>; 4]; 4] {
    let cc = ChargeCombinations { nf };
    const E2M: f64 = EU2 - ED2;

    let gamma_phu = gamma_phu(c, nf);
    let gamma_phd = gamma_phd(c, nf);
    let gamma_uph = gamma_uph(c, nf);
    let gamma_dph = gamma_dph(c, nf);
    let gamma_nspu = gamma_nspu(c, nf);
    let gamma_nspd = gamma_nspd(c, nf);
    let gamma_ps = gamma_ps(c, nf);

    [
        [
            Complex::<f64>::zero(),
            Complex::<f64>::zero(),
            Complex::<f64>::zero(),
            Complex::<f64>::zero(),
        ],
        [
            Complex::<f64>::zero(),
            gamma_phph(c, nf),
            cc.vu() * EU2 * gamma_phu + cc.vd() * ED2 * gamma_phd,
            cc.vu() * (EU2 * gamma_phu - ED2 * gamma_phd),
        ],
        [
            Complex::<f64>::zero(),
            cc.vu() * EU2 * gamma_uph + cc.vd() * ED2 * gamma_dph,
            cc.vu() * EU2 * gamma_nspu + cc.vd() * ED2 * gamma_nspd + cc.e2avg().powi(2) * gamma_ps,
            cc.vu() * (EU2 * gamma_nspu - ED2 * gamma_nspd + E2M * cc.e2avg() * gamma_ps),
        ],
        [
            Complex::<f64>::zero(),
            cc.vd() * (EU2 * gamma_uph - ED2 * gamma_dph),
            cc.vd() * (EU2 * gamma_nspu - ED2 * gamma_nspd + E2M * cc.e2avg() * gamma_ps),
            cc.vd() * EU2 * gamma_nspu
                + cc.vu() * ED2 * gamma_nspd
                + cc.vu() * cc.vd() * E2M.powi(2) * gamma_ps,
        ],
    ]
}

/// Compute the valence anomalous dimension matrix.
pub fn gamma_valence(c: &mut Cache, nf: u8) -> [[Complex<f64>; 2]; 2] {
    let cc = ChargeCombinations { nf };
    let gamma_nsmu = gamma_nsmu(c, nf);
    let gamma_nsmd = gamma_nsmd(c, nf);
    [
        [
            cc.vu() * EU2 * gamma_nsmu + cc.vd() * ED2 * gamma_nsmd,
            cc.vu() * (EU2 * gamma_nsmu - ED2 * gamma_nsmd),
        ],
        [
            cc.vd() * (EU2 * gamma_nsmu - ED2 * gamma_nsmd),
            cc.vd() * EU2 * gamma_nsmu + cc.vu() * ED2 * gamma_nsmd,
        ],
    ]
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{assert_approx_eq_cmplx, cmplx};
    use num::complex::Complex;
    use num::Zero;

    #[test]
    fn number_conservation() {
        const N: Complex<f64> = cmplx!(1., 0.);
        let mut c = Cache::new(N);
        for nf in 2..7 {
            assert_approx_eq_cmplx!(f64, gamma_nsmu(&mut c, nf), Complex::zero(), epsilon = 1e-5);
            assert_approx_eq_cmplx!(f64, gamma_nsmd(&mut c, nf), Complex::zero(), epsilon = 1e-5);
        }
    }

    #[test]
    fn quark_momentum_conservation() {
        const N: Complex<f64> = cmplx!(2., 0.);
        let mut c = Cache::new(N);
        for nf in 2..7 {
            let cc = ChargeCombinations { nf };
            let ps = EU2 * gamma_ps(&mut c, cc.nu()) + ED2 * gamma_ps(&mut c, cc.nd());
            let me_u = gamma_nspu(&mut c, nf) + ps + gamma_phu(&mut c, nf);
            assert_approx_eq_cmplx!(f64, me_u, Complex::zero(), epsilon = 1e-5);
            let me_d = gamma_nspd(&mut c, nf) + ps + gamma_phd(&mut c, nf);
            assert_approx_eq_cmplx!(f64, me_d, Complex::zero(), epsilon = 1e-5);
        }
    }

    #[test]
    fn photon_momentum_conservation() {
        const N: Complex<f64> = cmplx!(2., 0.);
        let mut c = Cache::new(N);

        for nf in 2..7 {
            let cc = ChargeCombinations { nf };
            assert_approx_eq_cmplx!(
                f64,
                EU2 * gamma_uph(&mut c, cc.nu())
                    + ED2 * gamma_dph(&mut c, cc.nd())
                    + gamma_phph(&mut c, nf),
                Complex::zero(),
                epsilon = 1e-12
            );
        }
    }
}
